File: electron/menu.ts
/**
 * Application menu — replaces Electron's default menu with a custom template
 * that mirrors the in-app Help sidebar.
 *
 * The interesting part is the Help menu: instead of the default "Search"
 * stub, we surface a Documentation submenu whose items are the same 13
 * topics rendered by src/pages/Help.tsx. Clicking a menu item sends an IPC
 * message 'help:navigate' with the topic id; the renderer routes to
 * /help?topic=<id> and Help.tsx reads the search param to select the topic.
 *
 * Topic ids MUST match those in src/pages/Help.tsx's TOPICS array. The
 * grouping mirrors the in-page sidebar (Start here / Setup / Workflows /
 * Sharing & export), separated by visual separators.
 *
 * All other menus are standard Electron roles so we don't reinvent
 * Undo/Redo/Reload/etc.
 */
import { app, Menu, BrowserWindow, shell, type MenuItemConstructorOptions } from 'electron'
import path from 'path'

// Topic id + label. Keep in sync with src/pages/Help.tsx TOPICS array — the
// id is what the renderer router uses to select the topic.
interface HelpTopic { id: string; label: string }

const HELP_TOPICS_START: HelpTopic[] = [
  { id: 'getting-started', label: 'Getting Started' },
]
const HELP_TOPICS_SETUP: HelpTopic[] = [
  { id: 'setup', label: 'Setup Tab' },
]
// Mirrors the workspace tab strip phases: Explore → Measure → Verify.
const HELP_TOPICS_WORKFLOWS: HelpTopic[] = [
  { id: 'coverage', label: 'Coverage' },
  { id: 'map',      label: 'Map' },
  { id: 'read',     label: 'Read' },
  { id: 'discover', label: 'Discover' },
  { id: 'score',    label: 'Score' },
  { id: 'track',    label: 'Track' },
  { id: 'compare',  label: 'Compare' },
  { id: 'audit',    label: 'Audit' },
  { id: 'gap',      label: 'Gap' },
]
const HELP_TOPICS_SHARING: HelpTopic[] = [
  { id: 'paper-bundle',   label: 'Paper-ready Bundle' },
  { id: 'project-bundle', label: 'Project Bundle (.lens)' },
]

/** Build a Documentation submenu item that fires `help:navigate` IPC on click. */
function helpTopicItem(window: BrowserWindow | null, topic: HelpTopic): MenuItemConstructorOptions {
  return {
    label: topic.label,
    click: () => {
      window?.webContents.send('help:navigate', topic.id)
    },
  }
}

/**
 * Resolve the user-manual PDF path for the current run mode.
 *  - packaged: copied to process.resourcesPath via electron-builder extraResources
 *  - dev:      repo root (sibling of electron/)
 *
 * Returns null if the file can't be located, so the menu can omit the item
 * cleanly rather than ship a broken link.
 */
function resolveUserManualPath(): string | null {
  const fs = require('fs') as typeof import('fs')
  const candidates = app.isPackaged
    ? [path.join(process.resourcesPath, 'Document-Lens-User-Manual.pdf')]
    : [
        path.join(__dirname, '..', 'Document-Lens-User-Manual.pdf'),
        path.join(__dirname, '..', '..', 'Document-Lens-User-Manual.pdf'),
      ]
  for (const p of candidates) {
    if (fs.existsSync(p)) return p
  }
  return null
}

/**
 * Build the full application menu template. Standard roles for File / Edit /
 * View / Window, custom Help. On macOS, prepends the App menu (about, prefs,
 * quit) — required for native feel.
 */
export function buildMenu(
  window: BrowserWindow | null,
  onCheckForUpdates?: () => void
): Menu {
  const isMac = process.platform === 'darwin'
  const manualPath = resolveUserManualPath()

  const checkForUpdatesItem: MenuItemConstructorOptions = {
    label: 'Check for Updates…',
    click: () => onCheckForUpdates?.(),
  }

  const documentationSubmenu: MenuItemConstructorOptions[] = [
    ...HELP_TOPICS_START.map((t) => helpTopicItem(window, t)),
    { type: 'separator' },
    ...HELP_TOPICS_SETUP.map((t) => helpTopicItem(window, t)),
    { type: 'separator' },
    ...HELP_TOPICS_WORKFLOWS.map((t) => helpTopicItem(window, t)),
    { type: 'separator' },
    ...HELP_TOPICS_SHARING.map((t) => helpTopicItem(window, t)),
  ]

  const helpSubmenu: MenuItemConstructorOptions[] = [
    {
      label: 'Documentation',
      submenu: documentationSubmenu,
    },
  ]
  if (manualPath) {
    helpSubmenu.push({ type: 'separator' })
    helpSubmenu.push({
      label: 'Open User Manual (PDF)',
      click: () => {
        shell.openPath(manualPath).catch((err) => console.error('Failed to open manual:', err))
      },
    })
  }
  if (!isMac) {
    // macOS surfaces About + Check for Updates inside the App menu (added
    // below). On Windows/Linux, the conventional place is the Help menu.
    helpSubmenu.push({ type: 'separator' })
    helpSubmenu.push(checkForUpdatesItem)
    helpSubmenu.push({ role: 'about' })
  }

  const template: MenuItemConstructorOptions[] = [
    // App menu — macOS only. Quit/Hide/etc. by convention live here on Mac;
    // on Windows/Linux Quit goes in the File menu (below).
    ...(isMac ? [{
      label: app.name,
      submenu: [
        { role: 'about' },
        checkForUpdatesItem,
        { type: 'separator' },
        { role: 'services' },
        { type: 'separator' },
        { role: 'hide' },
        { role: 'hideOthers' },
        { role: 'unhide' },
        { type: 'separator' },
        { role: 'quit' },
      ] as MenuItemConstructorOptions[],
    }] : []),

    // File menu — minimal; the app's project/file operations live in-UI.
    // We provide Close Window + Quit (non-mac) so the menu doesn't look empty.
    {
      label: 'File',
      submenu: [
        isMac ? { role: 'close' } : { role: 'quit' },
      ] as MenuItemConstructorOptions[],
    },

    // Edit — standard text-editing roles. The DB-management bits (find, etc.)
    // are in-UI on the Library page; the menu just covers OS-standard editing.
    {
      label: 'Edit',
      submenu: [
        { role: 'undo' },
        { role: 'redo' },
        { type: 'separator' },
        { role: 'cut' },
        { role: 'copy' },
        { role: 'paste' },
        ...(isMac
          ? [
              { role: 'pasteAndMatchStyle' as const },
              { role: 'delete' as const },
              { role: 'selectAll' as const },
              { type: 'separator' as const },
              {
                label: 'Speech',
                submenu: [
                  { role: 'startSpeaking' as const },
                  { role: 'stopSpeaking' as const },
                ],
              },
            ]
          : [
              { role: 'delete' as const },
              { type: 'separator' as const },
              { role: 'selectAll' as const },
            ]),
      ] as MenuItemConstructorOptions[],
    },

    // View — standard Chromium dev/zoom helpers. Reload + DevTools are
    // valuable for an Electron app, especially during user-reported bugs.
    {
      label: 'View',
      submenu: [
        { role: 'reload' },
        { role: 'forceReload' },
        { role: 'toggleDevTools' },
        { type: 'separator' },
        { role: 'resetZoom' },
        { role: 'zoomIn' },
        { role: 'zoomOut' },
        { type: 'separator' },
        { role: 'togglefullscreen' },
      ] as MenuItemConstructorOptions[],
    },

    // Window — minimize / close, plus the Mac-specific arrangement helpers.
    {
      label: 'Window',
      submenu: [
        { role: 'minimize' },
        { role: 'zoom' },
        ...(isMac
          ? [
              { type: 'separator' as const },
              { role: 'front' as const },
              { type: 'separator' as const },
              { role: 'window' as const },
            ]
          : [{ role: 'close' as const }]),
      ] as MenuItemConstructorOptions[],
    },

    // Help — the bespoke part.
    {
      role: 'help',
      submenu: helpSubmenu,
    },
  ]

  return Menu.buildFromTemplate(template)
}

File: src-tauri/src/menu.rs
//! Application menu — Rust port of electron/menu.ts.
//!
//! The bespoke part is Help → Documentation: the same 13 topics as the in-app
//! sidebar (src/pages/Help.tsx). Each item carries a `help:<topic>` id; on
//! click we emit `help:navigate` with the topic, and the renderer routes to
//! /help?topic=<id>. Topic ids MUST match Help.tsx's TOPICS array.
//!
//! Everything else uses Tauri predefined items (undo/redo/cut/copy/paste/…) so
//! we don't reinvent standard editing/window behaviour. "Open User Manual"
//! opens the bundled PDF; "Check for Updates" is wired in Phase 6.

use std::path::PathBuf;

use tauri::menu::{Menu, MenuBuilder, MenuEvent, SubmenuBuilder};
use tauri::{AppHandle, Emitter, Manager, Runtime};
use tauri_plugin_opener::OpenerExt;

// (id, label) — the `help:` prefix is added when building the item id.
// Grouped exactly like the in-page sidebar, separators between groups.
const START: &[(&str, &str)] = &[("getting-started", "Getting Started")];
const SETUP: &[(&str, &str)] = &[("setup", "Setup Tab")];
const WORKFLOWS: &[(&str, &str)] = &[
    ("coverage", "Coverage"),
    ("map", "Map"),
    ("read", "Read"),
    ("discover", "Discover"),
    ("score", "Score"),
    ("track", "Track"),
    ("compare", "Compare"),
    ("audit", "Audit"),
    ("gap", "Gap"),
];
const SHARING: &[(&str, &str)] = &[
    ("paper-bundle", "Paper-ready Bundle"),
    ("project-bundle", "Project Bundle (.lens)"),
];

const MANUAL_PDF: &str = "Document-Lens-User-Manual.pdf";

pub fn build<R: Runtime>(app: &AppHandle<R>) -> tauri::Result<Menu<R>> {
    let is_mac = cfg!(target_os = "macos");

    // Documentation submenu — the help topics, grouped with separators. Built
    // inline (rather than via a helper) so the builder's type is inferred.
    let mut doc = SubmenuBuilder::new(app, "Documentation");
    for (i, group) in [START, SETUP, WORKFLOWS, SHARING].iter().enumerate() {
        if i > 0 {
            doc = doc.separator();
        }
        for (id, label) in *group {
            doc = doc.text(format!("help:{id}"), *label);
        }
    }
    let doc = doc.build()?;

    // Help menu.
    let mut help = SubmenuBuilder::new(app, "Help").item(&doc).separator();
    help = help.text("help:manual", "Open User Manual (PDF)");
    if !is_mac {
        // macOS surfaces About + Check for Updates in the App menu instead.
        help = help
            .separator()
            .text("help:updates", "Check for Updates…")
            .about(None);
    }
    let help = help.build()?;

    let mut builder = MenuBuilder::new(app);

    // App menu — macOS only. Quit/Hide/etc. live here by convention on Mac.
    if is_mac {
        let app_menu = SubmenuBuilder::new(app, app.package_info().name.clone())
            .about(None)
            .text("help:updates", "Check for Updates…")
            .separator()
            .services()
            .separator()
            .hide()
            .hide_others()
            .show_all()
            .separator()
            .quit()
            .build()?;
        builder = builder.item(&app_menu);
    }

    // Edit — standard editing roles.
    let mut edit = SubmenuBuilder::new(app, "Edit")
        .undo()
        .redo()
        .separator()
        .cut()
        .copy()
        .paste();
    if is_mac {
        edit = edit.select_all();
    } else {
        edit = edit.separator().select_all();
    }
    let edit = edit.build()?;

    // Window — minimize / maximize / fullscreen / close.
    let window = SubmenuBuilder::new(app, "Window")
        .minimize()
        .maximize()
        .separator()
        .fullscreen()
        .separator()
        .close_window()
        .build()?;

    // File — minimal (the app's project/file ops are in-UI).
    let file = SubmenuBuilder::new(app, "File").close_window().build()?;

    builder = builder.item(&file).item(&edit).item(&window).item(&help);
    builder.build()
}

pub fn handle<R: Runtime>(app: &AppHandle<R>, event: MenuEvent) {
    let id = event.id().0.as_str();
    let Some(topic) = id.strip_prefix("help:") else {
        return;
    };
    match topic {
        "manual" => open_manual(app),
        "updates" => { /* wired in Phase 6 (updater) */ }
        // A documentation topic → route the renderer to /help?topic=<id>.
        _ => {
            let _ = app.emit("help:navigate", topic);
        }
    }
}

/// Locate the user-manual PDF. Prod: the bundled resource dir. Dev: walk up
/// from the cwd (the repo root holds the PDF).
fn resolve_manual<R: Runtime>(app: &AppHandle<R>) -> Option<PathBuf> {
    if let Ok(res) = app.path().resource_dir() {
        let p = res.join(MANUAL_PDF);
        if p.exists() {
            return Some(p);
        }
    }
    let cwd = std::env::current_dir().ok()?;
    for ancestor in cwd.ancestors() {
        let p = ancestor.join("document-lens").join(MANUAL_PDF);
        if p.exists() {
            return Some(p);
        }
        let p = ancestor.join(MANUAL_PDF);
        if p.exists() {
            return Some(p);
        }
    }
    None
}

fn open_manual<R: Runtime>(app: &AppHandle<R>) {
    match resolve_manual(app) {
        Some(path) => {
            if let Err(e) = app.opener().open_path(path.to_string_lossy().to_string(), None::<&str>) {
                eprintln!("[menu] failed to open user manual: {e}");
            }
        }
        None => eprintln!("[menu] user manual PDF not found"),
    }
}
